### PR TITLE
 Add Qwen3.5 MoE Model to Hub

### PR DESCRIPTION
**Pull request by laxmareddyp**
## Description of the change

This solves the feature request  issue #2632 

**This model extends the standard dense [Qwen3.5](https://github.com/keras-team/keras-hub/pull/2631) architecture by introducing a Mixture of Experts (MoE) design.**

**Qwen3.5 features the following enhancement:**

**Unified Vision-Language Foundation:**  Early fusion training on trillions of multimodal tokens achieves cross-generational parity with Qwen3 and outperforms Qwen3-VL models across reasoning, coding, agents, and visual understanding benchmarks.

**Efficient Hybrid Architecture:** Gated Delta Networks combined with sparse Mixture-of-Experts deliver high-throughput inference with minimal latency and cost overhead.

**Scalable RL Generalization:** Reinforcement learning scaled across million-agent environments with progressively complex task distributions for robust real-world adaptability.

**Global Linguistic Coverage:** Expanded support to 201 languages and dialects, enabling inclusive, worldwide deployment with nuanced cultural and regional understanding.

**Next-Generation Training Infrastructure:**  Near-100% multimodal training efficiency compared to text-only training and asynchronous RL frameworks supporting massive-scale agent scaffolds and environment orchestration.

## Numeric verification screenshots:

Screenshot 2026-07-02 at 6 40 30 PM
Screenshot 2026-07-02 at 6 41 06 PM

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

**Comment by laxmareddyp** [comment]
@divyashreepathihalli @sachinprasadhs 

Gemini review comments — Triaged all comments against the existing KerasHub codebase. 
* Three were addressed: moved import traceback to top-level (PEP 8), 
* Added # noqa: E501 for unavoidable long import paths, 
* Improved bool((i+1)%4) readability. 
* The remaining comments were not actioned because they flag patterns that are consistently used across the codebase — 
    * the batched-einsum MoE dispatch matches Gemma4 and Qwen3 MoE,
    *  if training: is standard in all MoE models (Keras 3 passes Python bools, not symbolic tensors), 
    * assert len(batch_shape) == 2 is used by 10+ CausalLM classes,
    *  float32 recurrent cache is required by GatedDeltaNet's numerical stability( Matches huggingFace config) 
    * the converter's layer_types default matches the dense Qwen3.5/Gemma4 converter pattern. 
* Changing these only in this model would introduce inconsistency.

I have opened up the gemini improve style guide PR , will add my observations to that PR to improve the gemini comments for subsequent fresh PR's.  

Thanks.

**Comment by laxmareddyp** [comment]
@divyashreepathihalli The GPU failure is unrelated to this model code changes; it is associated with `FNetEncoderTest. `All tests for this PR passed on an earlier commit, and the recent commit only resolved a conflict from the master branch.

**Comment by divyashreepathihalli** [comment]
hey can you add numerics screenshot?

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+    def call(self, hidden_states):
+        gate_up = ops.einsum(
+            "th,ehm->etm", hidden_states, self._expert_feedforward_gate_dense
+        )
+        gate, up = ops.split(gate_up, 2, axis=-1)
+        gate = ops.cast(gate, "float32")
+        gate = ops.silu(gate)
+        gate = ops.cast(gate, hidden_states.dtype)
+        hidden = up * gate
+        out = ops.einsum(
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:200]: The implementation computes the output of ALL experts for ALL tokens using ops.einsum. can we use a grouped GEMM or at least skip computation for experts with zero routed tokens if the backend supports it efficiently

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+        self._shared_expert_gate.build(decoder_sequence_shape)
+
+        self.built = True
+
+    def call(self, hidden_states, attention_mask=None, training=None):
+        batch_size, seq_len, _ = ops.shape(hidden_states)
+        hidden_states_flat = ops.reshape(hidden_states, (-1, self.hidden_dim))
+
+        shared_expert_output = self.shared_expert(hidden_states_flat)
+        shared_gate = ops.sigmoid(self._shared_expert_gate(hidden_states_flat))
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:289]: Is this sigmoid gate part of the official Qwen 3.5 MoE specification? Some models (like DeepSeek) use different fusion methods or no gating for the shared expert. We should verify this against the reference implementation.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+            * self.backbone.linear_value_head_dim
+        )
+        conv_dim = linear_key_dim * 2 + linear_val_dim
+        conv_shape = [
+            batch_size,
+            num_layers,
+            conv_dim,
+            self.backbone.linear_conv_kernel_dim - 1,
+        ]
+        conv_cache = ops.zeros(conv_shape, dtype=self.compute_dtype)
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py:161]: the cache logic looks complicated, we need to make sure it is well tested in tests files

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+        )
+        gate, up = ops.split(gate_up, 2, axis=-1)
+        gate = ops.cast(gate, "float32")
+        gate = ops.silu(gate)
+        gate = ops.cast(gate, hidden_states.dtype)
+        hidden = up * gate
+        out = ops.einsum(
+            "eti,eih->eth", hidden, self._expert_feedforward_output_dense
+        )
+        return out
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:203]: A truly sparse implementation is currently unfeasible in a backend-agnostic way because the Keras ops API lacks ragged_dot support. This matches the established architectural pattern used across all other MoE models in KerasHub

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+
+    Returns:
+        Scalar tensor representing the auxiliary loss.
+    """
+    routing_weights = ops.softmax(router_logits, axis=-1)
+    _, selected_experts = ops.top_k(routing_weights, k=top_k)
+    expert_mask = ops.one_hot(selected_experts, num_experts)
+
+    if attention_mask is not None:
+        batch_size, seq_len, _ = ops.shape(attention_mask)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py]: Valid and updated the code.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+            )
+            if self_attention_cache is not None:
+                x, self_attention_cache = x
+
+        x = self._self_attention_dropout(x, training=training)
+        x = x + residual
+
+        residual = x
+        x = self._post_attention_layernorm(x)
+        x = self.mlp(x, training=training)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py]: Valid and updated.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        if isinstance(token_ids, dict):
+            padding_mask = token_ids.get("padding_mask", padding_mask)
+            pixel_values = token_ids.get("pixel_values", None)
+            image_grid_thw = token_ids.get("image_grid_thw", None)
+            vision_indices = token_ids.get("vision_indices", None)
+            position_ids = token_ids.get("position_ids", None)
+            token_ids = token_ids["token_ids"]
+
+        batch_shape = ops.shape(token_ids)[:2]
+        assert len(batch_shape) == 2
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py]: Updated.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py (hunk 1 above)
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:200]: This is the established pattern across all MoE models in KerasHub (Mixtral, Qwen3MoE, QwenMoE, GptOss). Grouped GEMM / ragged_dot isn't available in keras.ops, and skipping zero-routed experts requires data-dependent control flow that breaks XLA compilation. Added a NOTE in the docstring documenting this design decision.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py (hunk 2 above)
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:289]: The sigmoid-gated shared expert is the official Qwen3/Qwen3.5 MoE design,
 this is verified against the HuggingFace Qwen3MoeSparseMoeBlock reference implementation and the checkpoint contains the shared_expert_gate weight.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py]: Duplicate comment

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_backbone.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+                shared_expert_intermediate_size=(
+                    shared_expert_intermediate_size
+                ),
+                num_experts=num_experts,
+                top_k=top_k,
+                partial_rotary_factor=partial_rotary_factor,
+                rope_max_wavelength=rope_max_wavelength,
+                rope_scaling_factor=rope_scaling_factor,
+                layer_norm_epsilon=layer_norm_epsilon,
+                activation=ops.silu,
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_backbone.py]: Updated

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+
+        if isinstance(x, tuple):
+            x, _ = x
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py]: Updated..!

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+        )
+
+        if training:
+            aux_loss = compute_load_balancing_loss(
+                router_logits=router_logits,
+                num_experts=self.num_experts,
+                top_k=self.top_k,
+                attention_mask=attention_mask,
+            )
+            self.add_loss(self.router_aux_loss_coefficient * aux_loss)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py:319]: this is the same pattern across all MoE models, changing it just for Qwen3.5 MoE would be inconsistent. The right approach is to keep consistency with the existing codebase.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py [hunk 11]
```diff
@@ -0,0 +1,659 @@
...
+    routing_weights = ops.softmax(router_logits, axis=-1)
+    _, selected_experts = ops.top_k(routing_weights, k=top_k)
+    expert_mask = ops.one_hot(selected_experts, num_experts)
+
+    if attention_mask is not None:
+        batch_size, seq_len, _ = ops.shape(attention_mask)
+        flat_mask = ops.any(attention_mask, axis=-1)
+        flat_mask = ops.reshape(flat_mask, (-1,))
+        expert_attention_mask = ops.expand_dims(flat_mask, axis=-1)
+        expert_attention_mask = ops.cast(expert_attention_mask, dtype="float32")
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_decoder.py]: Valid and updated.

File: keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py (hunk 3 above)
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5_moe/qwen3_5_moe_causal_lm.py:161]: The existing tests exercise the cache indirectly through test_generate, test_early_stopping, and test_generate_compilation — all of which call generate() which calls _build_cache() + call_with_cache(), Anyways added a dedicated tests.

**Comment by laxmareddyp** [comment]
> hey can you add numerics screenshot?

@divyashreepathihalli Added the numerics verification screenshots now.

**Review by divyashreepathihalli — APPROVED** [review]

**Comment by laxmareddyp** [comment]
@divyashreepathihalli These GPU test failures are unrelated to our recent changes. It looks like a flaky test caused by a missing tokenizer.json file in the Kaggle presets for the GPT-2 model. I will re-run the tests after some time. Thanks.
